File: IT/2/Multithreading/Esercizi/C++/esercizio5.cpp
/*
file esercizio5.cpp
compilare con: g++ -std=c++11 esercizio5.cpp -lpthread
*/

#include <iostream>
#include <random>
#include <future>

using std::cin;
using std::cout;
using std::endl;
using std::random_device;
using std::default_random_engine;
using std::uniform_int_distribution;
using std::async;
using std::future;

void func() {
	printf("Ciao, mondo!\n");
}

int main()
{
	default_random_engine gen(random_device{}());
	uniform_int_distribution<int> dist(2, 10);

	int nThreads = dist(gen);

	cout << "Verrà creato un pool di " << nThreads << " thread." << endl;

	for (int i = 0; i < nThreads; i++) {
		async(std::launch::async, func);
	}
}

File: IT/2/Multithreading/Esercizi/Python/esercizio5.py
# file esercizio5.py
# -*- coding: utf-8 -*-

from __future__ import print_function
from multiprocessing import Pool
from random import randint

def func():
	print("Ciao, mondo!")

if __name__ == "__main__":
	nProc = randint(2, 10)

	print("Verrà creato un pool di {} processi.".format(nProc))

	pool = Pool(processes=nProc)

	for i in range(nProc):
		pool.apply_async(func)

	pool.close()
	pool.join()

